File: Sources/ClipboardCore/ClipReceiver.swift
import Foundation

public enum ReceiveResult: Equatable, Sendable {
    case added(ClipItem)
    case duplicate
}

@MainActor
public final class ClipReceiver {
    private let store: ClipStore
    private let pasteboard: PasteboardClient

    public init(store: ClipStore, pasteboard: PasteboardClient) {
        self.store = store
        self.pasteboard = pasteboard
    }

    public func receive(_ payload: ClipPayload) throws -> ReceiveResult {
        let initialCount = store.items.count
        let addedClip: ClipItem

        switch payload.type {
        case .text:
            guard let text = payload.text else {
                throw NSError(domain: "ClipReceiver", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing text payload"])
            }
            addedClip = try store.addText(text)
            
            pasteboard.clearContents()
            pasteboard.writeText(text)
            
        case .image:
            guard let base64 = payload.imageBase64, let data = Data(base64Encoded: base64) else {
                throw NSError(domain: "ClipReceiver", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing or invalid image payload"])
            }
            let ext = payload.imageExtension ?? "png"
            addedClip = try store.addImageData(data, fileExtension: ext, previewText: payload.previewText)
            
            pasteboard.clearContents()
            pasteboard.writeImageData(data)
        }
        
        if initialCount == store.items.count {
            return .duplicate
        } else {
            return .added(addedClip)
        }
    }
}

File: ClipboardSS.Windows/src/ClipboardSS.Core/Sync/ClipReceiver.cs
using ClipboardSS.Core.Models;
using ClipboardSS.Core.Storage;

namespace ClipboardSS.Core.Sync;

public enum ReceiveStatus
{
    Added,
    Duplicate,
}

public sealed record ReceiveResult(ReceiveStatus Status, ClipItem? Clip = null);

public sealed class ClipReceiver(ClipStore store, IPasteboardClient pasteboard)
{
    public ReceiveResult Receive(ClipPayload payload)
    {
        var initialCount = store.Items.Count;
        ClipItem clip;
        switch (payload.Type)
        {
            case ClipType.Text:
                if (payload.Text is null)
                {
                    throw new InvalidDataException("Missing text payload.");
                }

                clip = store.AddText(payload.Text);
                pasteboard.ClearContents();
                pasteboard.WriteText(payload.Text);
                break;
            case ClipType.Image:
                if (payload.ImageBase64 is null)
                {
                    throw new InvalidDataException("Missing image payload.");
                }

                byte[] imageData;
                try
                {
                    imageData = Convert.FromBase64String(payload.ImageBase64);
                }
                catch (FormatException exception)
                {
                    throw new InvalidDataException("Invalid image payload.", exception);
                }

                clip = store.AddImageData(
                    imageData,
                    payload.ImageExtension ?? "png",
                    payload.PreviewText);
                pasteboard.ClearContents();
                pasteboard.WriteImageData(imageData);
                break;
            default:
                throw new InvalidDataException("Unknown clip type.");
        }

        return initialCount == store.Items.Count
            ? new ReceiveResult(ReceiveStatus.Duplicate)
            : new ReceiveResult(ReceiveStatus.Added, clip);
    }
}
